### DISCUSSION TITLE
Edit in Visual Mode changes the format of tables 🥺

### DISCUSSION THREAD

**Discussion by lilianavalenciar** [quarto-dev/quarto-cli#6405] [Q&A]
### Description

Hello there! 

Quarto version 1.2.475
macOS Ventura 13.4.1 (c) 

I use quarto for work, we have all our documentation in markdown format. I use Edit in Visual Mode to improve the format but mostly for tables. When working with tables, I'm having an issue. The table is in markdown and when I use Edit in Visual Mode and edit the table, this changes the format of the table to what I believe is html 🤯

[example](https://www.loom.com/share/f0b2af5f751d4f02857ec5a9662e570e?sid=a03b3413-cbcf-4237-9e15-800481aa6ea98)

It's extremely annoying and it's very recent. Help!

**Comment by mcanouil** [comment]
Would you mind sharing a reproducible example?
I currently can't reproduce with a simple table.

https://github.com/quarto-dev/quarto-cli/assets/8896044/0f912962-9a32-4415-aeeb-aa812866bd33

**Reply by lilianavalenciar** [reply]
I'm sorry, the error is not consistent and I don't really know what it's triggering it. I tried again with the same file I recorded in the loom video and now I'm not having this issue. However, my intuition is that this is triggered by special characters.

**Reply by mcanouil** [reply]
Pipe tables as in my example have to be simple, no multiline, etc.
I believe the Visual editor is switching to HTML when the table is no longer a valid pipe markdown table.

**Reply by lilianavalenciar** [reply]
Can be. Although, I have done more extensive tables in the past and this was the first time that I got that. Specially going from Visual Mode to Source Mode.  I'll keep and eye on it.

**Reply by lilianavalenciar** [reply]
I had the problem again and together with a colleague **we found out that Quarto doesn't like to have tables with more than 9 columns**.

**Reply by mcanouil** [reply]
Would you mind sharing a reproducible example (as plain text not as attachment)? (note that most of what is done by the Visual Editor, is actually done by Pandoc)

---

You can share a **self-contained "working" (reproducible)** Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).
See <https://quarto.org/bug-reports.html#small-is-beautiful-aim-for-a-single-document-with-10-lines>.

If you have multiple files (and if it is **absolutely required** to have multiple files), please share as a Git repository.

<table>
<tr><th>R</th><th>Python</th></tr>
<tr><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: knitr
---

This is a reproducible Quarto document.

```{r}
x <- c(1, 2, 3, 4, 5)
y <- c(1, 4, 9, 16, 25)

plot(x, y)
```

![An image](https://placehold.co/600x400.png)

The end.
````
`````

</td><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: jupyter
---

This is a reproducible Quarto document.

```{python}
import matplotlib.pyplot as plt

x = [1, 2, 3, 4, 5]
y = [1, 4, 9, 16, 25]

plt.plot(x, y)
plt.show()
```

![An image](https://placehold.co/600x400.png)

The end.
````
`````

</td></tr>
</table>

Additionally and if not already given, please share the output of `quarto check` within a code blocks (*i.e.*, using three backticks ` ```txt `), see <https://quarto.org/bug-reports.html#check>.

**Reply by lilianavalenciar** [reply]
To reproduce the error, use the following qmd. 

- Tables converted to HTML:
  - Go to Edit in Visual Mode and inside any cell of the table tap enter. 
  - Come back to source mode and voilà, the table is not HTML. 
- cmd+z to undo the changes in source mode
- Table preserved in markdown: 
   -  Go to Edit in Visual Mode
   -  Delete one column, tap enter inside any cell of the table. 
   - Come back to the source mode. The table is markdown! 


````qmd
---
title: Product risk assessment - FMEA
format: pdf
---


# Risk assessment

We perfomed a risk assessment for Usability (UFMEA), Design (DFMEA), Processes (PFMEA). We included the cybersecurity risk assessment (CFMEA) although we used the threat analysis instead of the hazard, hazardous sitatuation, harm principle.

The following notation is used in the tables:

P: Probablity

S: Severity

R: Risk level

| Risk ID  | Design Element                       | Hazardous situation                                                                                                             | Hazard                   | Sequence of events leading to harm                                       | Harm                                                    | Probability | Severity | Risk level | Risk control measure ID                                                                                                                     |
|----------|--------------------------------------|---------------------------------------------------------------------------------------------------------------------------------|--------------------------|--------------------------------------------------------------------------|---------------------------------------------------------|-------------|----------|------------|---------------------------------------------------------------------------------------------------------------------------------------------|
| DFMEA001 | Analysis worker                      | Failure of the Analysis worker prevent the user from accessing the caseThe user has to manually compare the scansAA             | aaaSoftware problem      | Device/system does not work temporarilyaaaaaaDelayed or manual reporting | No patient, user, or stakeholder harm                   | 3           | 1        | L          | REQ-072 Instructions to contact support in IFUREQ-081 Error Detection - Monitoring and Logging aaaaaaREQ-041 Analysis status                |
| DFMEA002 | Analysis worker                      | The analysis worker incorrectly detects lesions                                                                                 | Software problem         | Incorrect information displayedaaaaaaIncorrect reporting                 | Unnecessary treatment side effects / Worsened prognosis | 5           | 2        | M          | REQ-040 Validate newly appearing analysisaaaaaaREQ-063 Visual cues for lesion detection and data clarityaaaaaaREQ-066 Validation correction |
| DFMEA003 | DICOM Orchestrator / Analysis Worker | Unsupported DICOM is uploaded (e.g. high spacing, non-uniform spacing, incorrect modality)                                      | Inadequate data handling | Incorrect information displayedaaaaaaDelayed or manual reporting         | No patient, user, or stakeholder harm                   | 5           | 1        | L          | REQ-083 Image Data Protocol in IFUaaaaaaREQ-017 Discard Input Unsupported Data                                                              |
| DFMEA004 | DICOM Orchestrator                   | The network ports of the hospital network is incorrectly configured.aaaaaaThe system is unable to transfer the DICOM image data | Software problem         | Device/system does not work temporarilyaaaaaaDelayed or manual reporting | No patient, user, or stakeholder harm                   | 4           | 1        | L          | REQ-088 Technical description                                                                                                               |
| DFMEA005 | DICOM Orchestrator / Analysis Worker | Data from two patients are mixed-up                                                                                             | Inadequate data handling | Incorrect information displayedaaaaaaIncorrect reporting                 | Unnecessary treatment side effects / Worsened prognosis | 3           | 2        | L          | REQ-037 Patient Identification in Viewer                                                                                                    |
| DFMEA006 | Web-based UI                         | The browser crashes, the user doesn't finish validation                                                                         | Software problem         | Device/system does not work temporarilyaaaaaaDelayed or manual reporting | No patient, user, or stakeholder harm                   | 4           | 1        | L          | REQ-040 Validate newly appearing analysisaaaaaaREQ-066 Validation correction                                                                |
| DFMEA007 | Web-based UI                         | The web-based UI loses the connection with the backend                                                                          | Software problem         | Device/system does not work temporarilyaaaaaaDelayed or manual reporting | No patient, user, or stakeholder harm                   | 4           | 1        | L          | REQ-040 Validate newly appearing analysisaaaaaaREQ-080 Error Handling for Failed Upload Validation                                          |
| DFMEA008 | Web-based UI                         | The web-based UI is loaded on an unsupported browser (e.g. IE10)                                                                | Software problem         | Device/system does not work temporarilyaaaaaaDelayed or manual reporting | No patient, user, or stakeholder harm                   | 3           | 1        | L          | REQ-073 Browser Compatibility Information in IFUaaaaaaREQ-074 Non-supported browser alert                                                   |
| DFMEA009 | Web-based UI                         | Slow internet connection                                                                                                        | Software problem         | Device/system does not work temporarilyaaaaaaDelayed or manual reporting | No patient, user, or stakeholder harm                   | 4           | 1        | L          | REQ-040 Validate newly appearing analysis                                                                                                   |
| DFMEA010 | Web-based UI                         | Firewalls                                                                                                                       | Software problem         | Device/system does not work temporarilyaaaaaaDelayed or manual reporting | No patient, user, or stakeholder harm                   | 4           | 1        | L          | REQ-040 Validate newly appearing analysisaaaaaaREQ-087 Firewall Warning in IFU                                                              |
| DFMEA012 | Analysis worker                      | An error in the registration causes all lesions to be considered new lesions                                                    | Software problem         | Incorrect information displayedaaaaaaIncorrect reporting                 | Unnecessary treatment side effects / Worsened prognosis | 3           | 2        | L          | REQ-083 Image Data Protocol in IFU                                                                                                          |

: Caption

````

````txt
---
Quarto 1.4.551
[✓] Checking versions of quarto binary dependencies...
      Pandoc version 3.1.11: OK
      Dart Sass version 1.69.5: OK
      Deno version 1.37.2: OK
[✓] Checking versions of quarto dependencies......OK
[✓] Checking Quarto installation......OK
      Version: 1.4.551
      Path: /Applications/quarto/bin

[✓] Checking tools....................OK
      TinyTeX: v2023.04
      Chromium: (not installed)

[✓] Checking LaTeX....................OK
      Using: TinyTex
      Path: /Users/lvalencia/Library/TinyTeX/bin/universal-darwin
      Version: 2023

[✓] Checking basic markdown render....OK

WARNING: Specified QUARTO_PYTHON '/Users/lvalencia/Documents/GitHub/AH-1/.venv/bin/python' does not exist.
[✓] Checking Python 3 installation....OK
      Version: 3.12.2
      Path: /opt/homebrew/opt/python@3.12/bin/python3.12
      Jupyter: (None)

      Jupyter is not available in this Python installation.
      Install with python3 -m pip install jupyter

      There is an unactivated Python environment in .venv. Did you forget to activate it?

[✓] Checking R installation...........OK
      Version: 4.2.3
      Path: /Library/Frameworks/R.framework/Resources
      LibPaths:
        - /Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/library
      knitr: 1.42
      rmarkdown: (None)

      The rmarkdown package is not available in this R installation.
      Install with install.packages("rmarkdown")
---
````

**Reply by mcanouil** [reply]
Is it on purpose that the grid is not aligned? (second row in particular)

**Reply by lilianavalenciar** [reply]
Not at all, I just copy paste. I work mostly on Visual Mode so I don't mind how the raw markdown table looks. I fixed for you ☺️

**Reply by mcanouil** [reply]
It's not for stylistic consideration, depending on the type of Markdown table, the alignment matters for the table to be properly parsed/displayed.

**Reply by cderv** [reply]
@mcanouil to be clear, the markdown shared is the one produced by Visual Editor itself for the pipe table. I don't think alignment of `|` is the issue here as Visual Mode parse table correctly.

**Reply by mcanouil** [reply]
@cderv I've got that part from https://github.com/quarto-dev/quarto-cli/discussions/6405#discussioncomment-8852687

I never said it was the issue as pipe table does not require alignment as far as I know, but grid tables do.

I asked to know the context of the creation of the table.

**Comment by cderv** [comment]
I don't think we document really how visual editor will parse and write tables. 

My understanding is that we prefer pipe_tables (https://pandoc.org/MANUAL.html#extension-pipe_tables) but they have this limitation
> The cells of pipe tables cannot contain block elements like paragraphs and lists, and cannot span multiple lines.

In the example you are doing 

> inside any cell of the table tap enter

this means you are inserting content in cell that is not support by pipe table (which is the syntax initially used in your source document)

In that case, I believe  Visual Editor should then consider `grid_tables` but somehow, the raw HTML is preferred, after a certain number of columns. If you remove one column in the example shared, then you'll have a grid table in the source markdown. I believe this is related to the `columns` information as mentioned in `tbl-colwidth` (https://quarto.org/docs/reference/formats/html.html#tables) 

The HTML representation is the one that allows to handle width and HTML table are passed anyway, so the resulting document will have the correct table. 


> It's extremely annoying and it's very recent.

Regarding the recent, note that you are using an old version of Quarto (1.2 and now 1.4 is the latest). The visual editor is part of IDE tooling (from https://github.com/quarto-dev/quarto) shared in VSCODE and RSTUDIO IDE. Not sure which one you are using. You should make sure to use a recent version also. 

So maybe there is an issue in how the content is detected, but it is possible that for complex wide table, the HTML representation is the only one that insures to get the correct output. 

Hope this helps clarify; 

Also, for the future, for issues on visual editor, you can open one at https://github.com/quarto-dev/quarto

**Reply by lilianavalenciar** [reply]
Hey there! 

Thanks for taking the time for this. 

If you check the quarto check I shared, I'm currently using the 1.4 version, although I run into the same issue when I was working with the 1.2.

> The HTML representation is the one that allows to handle width and HTML table are passed anyway, so the resulting document will have the correct table.

The problem with this is that after a couple iterations between Visual and Source mode, the content change to something unreadable, let's say. 

I work for a company developing Software a Medical Device and all our documentation is written in markdown/quarto. We have several tables and we're able to render tables with list and multiple lines. 

All that just to say that we manage to get most of the things we want to do and give you guys the information. We solved our issue with our docs for now. 

Cheers!

**Reply by cderv** [reply]
> If you check the quarto check I shared, I'm currently using the 1.4 version, although I run into the same issue when I was working with the 1.2.

Oh  I missed that sorry. 

> All that just to say that we manage to get most of the things we want to do and give you guys the information. We solved our issue with our docs for now.

Thanks a lot for the feedback. We'll pay attention to other report on this behavior and try to check if something is off in the heuristic for writing the right table format in Mardown based on what is added in the visual mode.
